DevCycle.SDK.Server.Local/Api/EventQueue.cs

```diff
@@ -62,127 +68,92 @@ public void RemoveFlushedEventsSubscriber(EventHandler<DVCEventArgs> flushedEven
+            logger.LogDebug($"DVC Flush ${eventCount} AS Events, for ${flushPayloads.Count} Users");
 
+            var requestTasks = flushPayloads.Select(GetPayloadResult).ToList();
+            await Task.WhenAll(requestTasks);
+            var results = requestTasks.Select(task => task.Result);
+            foreach (var (flushPayload, res) in results)
+            {
                 try
                 {
-                    foreach (var batch in batchQueue)
-                    {
-                        response = await dvcEventsApiClient.PublishEvents(batch);
-
-                        if (response.StatusCode == HttpStatusCode.Created) continue;
-                        var error = new DVCException(response.StatusCode,
-                            new ErrorResponse(response.Content ?? "Something went wrong flushing events"));
-
-                        if (!error.IsRetryable())
-                        {
-                            // Add non-retryable payloads to list of "completed" so that they are removed from queue
-                            completedRequests.Add(batch);
-                        }
-
-                        throw error;
-                    }
-
-                    batchQueue.Clear();
-
-                    logger.LogDebug("DVC Flushed {EventCount} Events, for {UserEventBatch} Users", eventCount,
-                        userEventBatch.Count);
-                    eventArgs.Success = true;
-                    OnFlushedEvents(eventArgs);
-                }
-                catch (DVCException e)
-                {
-                    if (e.IsRetryable())
+                    if (res.StatusCode != HttpStatusCode.Created)
                     {
-                        logger.LogError(
-                            "Error publishing events, retrying, status {status}, body: {response}", e.HttpStatusCode,
-                            response?.Content);
-                        ScheduleFlushWithDelay(true);
+                        logger.LogError($"Error publishing events, status: ${res.StatusCode}, body: ${res.Content}");
+                        localBucketing.OnPayloadFailure(this.environmentKey, flushPayload.PayloadID, (int)res.StatusCode >= 500);
+                        flushResultEvent.Success = false;
+                        flushResultEvent.Error = new DVCException(res.StatusCode, new ErrorResponse(res.ErrorMessage ?? ""));
                     }
                     else
                     {
-                        logger.LogError(
-                            "DVC Events were invalid and have been dropped, status {status}, body: {response}",
-                            e.HttpStatusCode,
-                            response?.Content);
+                        logger.LogDebug($"DVC Flushed ${eventCount} Events, for ${flushPayload.Records.Count} Users");
+                        localBucketing.OnPayloadSuccess(this.environmentKey, flushPayload.PayloadID);
```

Comment:
This call will need to be semaphore/mutex'd so we don't call localbucketing from more than once place since this is an async function.